saged/models.py

```diff
@@ -506,3 +508,180 @@ def get_parameters(self) -> Iterable[torch.Tensor]:
     def load_parameters(self, parameters: dict) -> "PytorchSupervised":
         self.model.load_state_dict(parameters)
         return self
+
+
+class UnsupervisedModel():
+    """
+    A model API defining the behavior of unsupervised models. Largely follows the sklearn model api
+    """
+    def __init__(self) -> None:
+        """
+        Standard model init function. We use pass instead of raising a NotImplementedError
+        here in case inheriting classes decide to call `super()`
+        """
+        pass
+
+    @abstractmethod
+    def load_model(classobject, model_path):
+        """
+        Read a pickeled model from a file and return it
+
+        Arguments
+        ---------
+        model_path: The location where the model is stored
+
+        Returns
+        -------
+        model: The model saved at `model_path`
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def fit(self, dataset: UnlabeledDataset) -> "UnsupervisedModel":
+        """
+        Train a model using the given unlabeled data
+
+        Arguments
+        ---------
+        dataset: The labeled data for use in training
+
+        Returns
+        -------
+        self: The trained version of the model
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def transform(self, dataset: UnlabeledDataset) -> UnlabeledDataset:
+        """
+        Use the learned embedding from the model to embed the given dataset
+
+        Arguments
+        ---------
+        dataset: The unlabeled data whose labels should be predicted
+
+        Returns
+        -------
+        predictions: A numpy array of predictions
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def save_model(self, out_path: str) -> None:
+        """
+        Write the model to a file
+
+        Arguments
+        ---------
+        out_path: The path to the file to write the classifier to
+
+        Raises
+        ------
+        FileNotFoundError if out_path isn't openable
+        """
+        raise NotImplementedError
+
+    def fit_transform(self, dataset: UnlabeledDataset) -> UnlabeledDataset:
```

Comment:
I think in the case of PCA this might be slower than just calling `model.fit_transform`, since your implementation has to call `get_all_data` twice which transposes the data matrix twice. I could be wrong, though - I don't have a good idea of how slow it is to take a transpose in Numpy (or Pandas).

Reply:
That's a good point, I didn't think about that. Apparently [it's a constant time operation](https://stackoverflow.com/questions/58279082/time-complexity-of-numpy-transpose). A point of interesting numpy trivia that I learned: the transpose operation is done by changing the way that functions access the underlying memory, so it doesn't actually modify the matrix at all

Reply:
Interesting! I didn't know that. In that case what you have should be fine.